runner/src/internal.rs

```diff
@@ -97,24 +97,76 @@ pub struct BuildServer {
     pub coverage: bool,
 }
 
+#[derive(Clone)]
+/// A construct to keep track of the status of the execution. It only cares about the top-level
+/// steps.
+pub struct Status {
+    failed: usize,
+    passed: usize,
+    remaining: usize,
+}
+
+impl Status {
+    pub fn new(remaining: usize) -> Self {
+        Status {
+            failed: 0,
+            passed: 0,
+            remaining,
+        }
+    }
+
+    /// Guarantees that the `failed`, `passed`, and `remaining` counts are updated only after the
+    /// completions of each top-level step.
+    fn update(&mut self, context: &Context, step_failed: bool) {
```

Comment:
This seems essentially another way of keeping track of the result from each compound step, which is already done in part by 

https://github.com/project-oak/oak/blob/9eb2b43e9b415c203943ada57da1a455a793ea9a/runner/src/internal.rs#L168

I think a more elegant way of doing this would be to compute the top-level list of futures, and check how many have already been resolved, and to what value? I think this makes sense especially seeing that this struct does not do anything apart from the top-level context, but it still needs to be passed around, which seems unnecessary

Reply:
I like that idea. :) If the futures already give us the information of which steps ran & and with which result it makes a lot of sense to use that, instead of separately keeping track of that status. :)